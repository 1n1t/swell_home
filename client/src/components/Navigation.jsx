/* eslint-disable jsx-a11y/anchor-is-valid,import/no-extraneous-dependencies,react/forbid-prop-types */
import React, { Component } from 'react';
import { withRouter } from 'react-router-dom';
import PropTypes from 'prop-types';

import NavigationLink from './NavigationLink';
import navData from '../data/navigation.json';
import smData from '../data/social-media.json';

class Navigation extends Component {
	static renderSocialMediaLinks() {
  	return smData.map(({url, cssClass}) => (
			<li className="link" key={url}>
				<a
					href={url}
					target="_blank"
					rel="noopener noreferrer"
				>
					<span className={cssClass} />
				</a>
			</li>
		))
	}

	state = {
		overlayColor: ''
	};

	componentDidMount() {
		this.setDefaultOverlayColor();
	}

	setDefaultOverlayColor() {
		const color = this.getDefaultOverlayColor();
		this.setState({overlayColor: color});
	};

	getDefaultOverlayColor = () => navData.filter(({ path }) => path === this.props.location.pathname)[0].overlayColor;

	changeOverlayColorHandler = (color = '') => {
		const overlayColor = (color === '') ? this.getDefaultOverlayColor() : color;
		this.setState({overlayColor});
	};

	renderNavLinks = () => navData.map(item =>
			<NavigationLink
				key={item.path}
				{...item}
				changeOverlayColor={this.changeOverlayColorHandler}
			/>);

	render() {
    const overlayStyle = {
      backgroundColor: this.state.overlayColor
    };

		return (
      <div className="Navigation">
        <nav className="navbar">
          <ul className="links">{this.renderNavLinks()}</ul>
          <ul className="social-media">{Navigation.renderSocialMediaLinks()}</ul>
        </nav>
        <div className="overlay" style={overlayStyle} />
      </div>
    );
  }
}

Navigation.propTypes = {
  location: PropTypes.object.isRequired
};

export default withRouter(Navigation);
