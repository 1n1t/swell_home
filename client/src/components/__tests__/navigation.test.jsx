import { Navigation } from '../Navigation';
import navLinks from '../../data/navigation.json';
import smLinks from '../../data/social-media.json';
import { sel, TestHelper } from '../../../../tests/utils';

const defaultProps = {
	location: {
		pathname: '/'
	}
};

const helper = new TestHelper(Navigation, defaultProps);

test('the component mounts with defaults', () => {
	const wrapper = helper.mountComponent({}, true);

	expect(wrapper).toMatchSnapshot();
});

test('it renders correct number of navigation and social media links', () => {
	const wrapper = helper.mountComponent();
	const elNavLinks = wrapper.find(sel('nav-links')).children();
	const elSocialMediaLinks = wrapper.find(sel('social-media-links')).children();

	expect(elNavLinks).toHaveLength(navLinks.length);
	expect(elSocialMediaLinks).toHaveLength(smLinks.length);
});

test('method "getDefaultOverlayColor()" returns correct color', () => {
	const pathname = '/contact';
	const wrapper = helper.mountComponent({ location: { pathname } }, true);
	const color = wrapper.instance().getDefaultOverlayColor();

	expect(color).toBe(navLinks.filter(item => item.path === pathname)[0].overlayColor);
});

xtest('it changes overlay color when user hover over a navigation link', () => {
	const index = 1;
	const wrapper = helper.mountComponent();
	const linkToHover = wrapper
		.find(sel('nav-link'))
		.children()
		.at(index);

	linkToHover.simulate('mouseenter');
	let overlayColor = wrapper.find(sel('overlay')).get(0).props.style.backgroundColor;
	expect(overlayColor).toBe(navLinks[index].overlayColor);

	linkToHover.simulate('mouseleave');
	overlayColor = wrapper.find(sel('overlay')).get(0).props.style.backgroundColor;
	expect(overlayColor).toBe(navLinks[0].overlayColor);
});

test('method "changeOverlayColorHandler()" changes overlay color', () => {
	const wrapper = helper.mountComponent({}, true);
	const backgroundColor = wrapper.instance().getDefaultOverlayColor();
	const newColor = '#001122';

	//	if color is provided, we expect overlay to have new color
	wrapper.instance().changeOverlayColorHandler(newColor);
	wrapper.update();
	let overlayStyle = wrapper.find(sel('overlay')).get(0).props.style;
	expect(overlayStyle).toEqual({ backgroundColor: newColor });

	//	if color is not provided, we expect overlay to have default color
	wrapper.instance().changeOverlayColorHandler();
	wrapper.update();
	overlayStyle = wrapper.find(sel('overlay')).get(0).props.style;
	expect(overlayStyle).toEqual({ backgroundColor });
});

test('it switch "is-open" class when toggle clicked', () => {
	const wrapper = helper.mountComponent();
	const toggle = wrapper.find(sel('nav-toggle'));

	toggle.simulate('click');
	expect(wrapper.find(sel('navigation')).hasClass('is-open')).toBe(true);

	toggle.simulate('click');
	expect(wrapper.find(sel('navigation')).hasClass('is-open')).toBe(false);
});
