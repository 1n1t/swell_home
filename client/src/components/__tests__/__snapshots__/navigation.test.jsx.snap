// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`the component mounts with defaults 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Navigation
    location={
      Object {
        "pathname": "/",
      }
    }
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <NavigationToggle
          onClickCallback={[Function]}
          toggledOn={false}
        />,
        <nav
          className="navbar"
        >
          <ul
            className="links"
            data-test="nav-links"
          >
            <NavigationLink
              changeOverlayColor={[Function]}
              onClickCallback={[Function]}
              overlayColor="#a9b7b4"
              path="/"
              text="Lookbook"
            />
            <NavigationLink
              changeOverlayColor={[Function]}
              onClickCallback={[Function]}
              overlayColor="#ec6761"
              path="/shop"
              text="Shop"
            />
            <NavigationLink
              changeOverlayColor={[Function]}
              onClickCallback={[Function]}
              overlayColor="#f7bfb1"
              path="/contact"
              text="Contact"
            />
            <NavigationLink
              changeOverlayColor={[Function]}
              onClickCallback={[Function]}
              overlayColor="#f9ca5c"
              path="/about"
              text="About"
            />
          </ul>
          <SocialMedis />
        </nav>,
        <div
          className="overlay"
          data-test="overlay"
          style={
            Object {
              "backgroundColor": "#a9b7b4",
            }
          }
        />,
      ],
      "className": "Navigation",
      "data-test": "navigation",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "onClickCallback": [Function],
          "toggledOn": false,
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <ul
              className="links"
              data-test="nav-links"
            >
              <NavigationLink
                changeOverlayColor={[Function]}
                onClickCallback={[Function]}
                overlayColor="#a9b7b4"
                path="/"
                text="Lookbook"
              />
              <NavigationLink
                changeOverlayColor={[Function]}
                onClickCallback={[Function]}
                overlayColor="#ec6761"
                path="/shop"
                text="Shop"
              />
              <NavigationLink
                changeOverlayColor={[Function]}
                onClickCallback={[Function]}
                overlayColor="#f7bfb1"
                path="/contact"
                text="Contact"
              />
              <NavigationLink
                changeOverlayColor={[Function]}
                onClickCallback={[Function]}
                overlayColor="#f9ca5c"
                path="/about"
                text="About"
              />
            </ul>,
            <SocialMedis />,
          ],
          "className": "navbar",
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": Array [
                <NavigationLink
                  changeOverlayColor={[Function]}
                  onClickCallback={[Function]}
                  overlayColor="#a9b7b4"
                  path="/"
                  text="Lookbook"
                />,
                <NavigationLink
                  changeOverlayColor={[Function]}
                  onClickCallback={[Function]}
                  overlayColor="#ec6761"
                  path="/shop"
                  text="Shop"
                />,
                <NavigationLink
                  changeOverlayColor={[Function]}
                  onClickCallback={[Function]}
                  overlayColor="#f7bfb1"
                  path="/contact"
                  text="Contact"
                />,
                <NavigationLink
                  changeOverlayColor={[Function]}
                  onClickCallback={[Function]}
                  overlayColor="#f9ca5c"
                  path="/about"
                  text="About"
                />,
              ],
              "className": "links",
              "data-test": "nav-links",
            },
            "ref": null,
            "rendered": Array [
              Object {
                "instance": null,
                "key": "/",
                "nodeType": "class",
                "props": Object {
                  "changeOverlayColor": [Function],
                  "onClickCallback": [Function],
                  "overlayColor": "#a9b7b4",
                  "path": "/",
                  "text": "Lookbook",
                },
                "ref": null,
                "rendered": null,
                "type": [Function],
              },
              Object {
                "instance": null,
                "key": "/shop",
                "nodeType": "class",
                "props": Object {
                  "changeOverlayColor": [Function],
                  "onClickCallback": [Function],
                  "overlayColor": "#ec6761",
                  "path": "/shop",
                  "text": "Shop",
                },
                "ref": null,
                "rendered": null,
                "type": [Function],
              },
              Object {
                "instance": null,
                "key": "/contact",
                "nodeType": "class",
                "props": Object {
                  "changeOverlayColor": [Function],
                  "onClickCallback": [Function],
                  "overlayColor": "#f7bfb1",
                  "path": "/contact",
                  "text": "Contact",
                },
                "ref": null,
                "rendered": null,
                "type": [Function],
              },
              Object {
                "instance": null,
                "key": "/about",
                "nodeType": "class",
                "props": Object {
                  "changeOverlayColor": [Function],
                  "onClickCallback": [Function],
                  "overlayColor": "#f9ca5c",
                  "path": "/about",
                  "text": "About",
                },
                "ref": null,
                "rendered": null,
                "type": [Function],
              },
            ],
            "type": "ul",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "function",
            "props": Object {},
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
        ],
        "type": "nav",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "className": "overlay",
          "data-test": "overlay",
          "style": Object {
            "backgroundColor": "#a9b7b4",
          },
        },
        "ref": null,
        "rendered": null,
        "type": "div",
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <NavigationToggle
            onClickCallback={[Function]}
            toggledOn={false}
          />,
          <nav
            className="navbar"
          >
            <ul
              className="links"
              data-test="nav-links"
            >
              <NavigationLink
                changeOverlayColor={[Function]}
                onClickCallback={[Function]}
                overlayColor="#a9b7b4"
                path="/"
                text="Lookbook"
              />
              <NavigationLink
                changeOverlayColor={[Function]}
                onClickCallback={[Function]}
                overlayColor="#ec6761"
                path="/shop"
                text="Shop"
              />
              <NavigationLink
                changeOverlayColor={[Function]}
                onClickCallback={[Function]}
                overlayColor="#f7bfb1"
                path="/contact"
                text="Contact"
              />
              <NavigationLink
                changeOverlayColor={[Function]}
                onClickCallback={[Function]}
                overlayColor="#f9ca5c"
                path="/about"
                text="About"
              />
            </ul>
            <SocialMedis />
          </nav>,
          <div
            className="overlay"
            data-test="overlay"
            style={
              Object {
                "backgroundColor": "#a9b7b4",
              }
            }
          />,
        ],
        "className": "Navigation",
        "data-test": "navigation",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "onClickCallback": [Function],
            "toggledOn": false,
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <ul
                className="links"
                data-test="nav-links"
              >
                <NavigationLink
                  changeOverlayColor={[Function]}
                  onClickCallback={[Function]}
                  overlayColor="#a9b7b4"
                  path="/"
                  text="Lookbook"
                />
                <NavigationLink
                  changeOverlayColor={[Function]}
                  onClickCallback={[Function]}
                  overlayColor="#ec6761"
                  path="/shop"
                  text="Shop"
                />
                <NavigationLink
                  changeOverlayColor={[Function]}
                  onClickCallback={[Function]}
                  overlayColor="#f7bfb1"
                  path="/contact"
                  text="Contact"
                />
                <NavigationLink
                  changeOverlayColor={[Function]}
                  onClickCallback={[Function]}
                  overlayColor="#f9ca5c"
                  path="/about"
                  text="About"
                />
              </ul>,
              <SocialMedis />,
            ],
            "className": "navbar",
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": Array [
                  <NavigationLink
                    changeOverlayColor={[Function]}
                    onClickCallback={[Function]}
                    overlayColor="#a9b7b4"
                    path="/"
                    text="Lookbook"
                  />,
                  <NavigationLink
                    changeOverlayColor={[Function]}
                    onClickCallback={[Function]}
                    overlayColor="#ec6761"
                    path="/shop"
                    text="Shop"
                  />,
                  <NavigationLink
                    changeOverlayColor={[Function]}
                    onClickCallback={[Function]}
                    overlayColor="#f7bfb1"
                    path="/contact"
                    text="Contact"
                  />,
                  <NavigationLink
                    changeOverlayColor={[Function]}
                    onClickCallback={[Function]}
                    overlayColor="#f9ca5c"
                    path="/about"
                    text="About"
                  />,
                ],
                "className": "links",
                "data-test": "nav-links",
              },
              "ref": null,
              "rendered": Array [
                Object {
                  "instance": null,
                  "key": "/",
                  "nodeType": "class",
                  "props": Object {
                    "changeOverlayColor": [Function],
                    "onClickCallback": [Function],
                    "overlayColor": "#a9b7b4",
                    "path": "/",
                    "text": "Lookbook",
                  },
                  "ref": null,
                  "rendered": null,
                  "type": [Function],
                },
                Object {
                  "instance": null,
                  "key": "/shop",
                  "nodeType": "class",
                  "props": Object {
                    "changeOverlayColor": [Function],
                    "onClickCallback": [Function],
                    "overlayColor": "#ec6761",
                    "path": "/shop",
                    "text": "Shop",
                  },
                  "ref": null,
                  "rendered": null,
                  "type": [Function],
                },
                Object {
                  "instance": null,
                  "key": "/contact",
                  "nodeType": "class",
                  "props": Object {
                    "changeOverlayColor": [Function],
                    "onClickCallback": [Function],
                    "overlayColor": "#f7bfb1",
                    "path": "/contact",
                    "text": "Contact",
                  },
                  "ref": null,
                  "rendered": null,
                  "type": [Function],
                },
                Object {
                  "instance": null,
                  "key": "/about",
                  "nodeType": "class",
                  "props": Object {
                    "changeOverlayColor": [Function],
                    "onClickCallback": [Function],
                    "overlayColor": "#f9ca5c",
                    "path": "/about",
                    "text": "About",
                  },
                  "ref": null,
                  "rendered": null,
                  "type": [Function],
                },
              ],
              "type": "ul",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "function",
              "props": Object {},
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
          ],
          "type": "nav",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "className": "overlay",
            "data-test": "overlay",
            "style": Object {
              "backgroundColor": "#a9b7b4",
            },
          },
          "ref": null,
          "rendered": null,
          "type": "div",
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
